utils/s3/client.cc

```diff
@@ -113,6 +114,230 @@ future<> client::delete_object(sstring object_name) {
     co_await _http.make_request(std::move(req), ignore_reply, http::reply::status_type::no_content);
 }
 
+class client::upload_sink : public data_sink_impl {
+    // "Each part must be at least 5 MB in size, except the last part."
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+    static constexpr size_t minimum_part_size = 5 << 20;
+    static constexpr int flush_concurrency = 3;
+
+    shared_ptr<client> _client;
+    http::experimental::client& _http;
+    sstring _object_name;
+    memory_data_sink_buffers _bufs;
+    sstring _upload_id;
+    utils::chunked_vector<sstring> _part_etags;
+    semaphore _flush_sem{flush_concurrency};
+
+    future<> start_upload();
+    future<> finalize_upload();
+    future<> maybe_flush();
+    future<> do_flush();
+    future<> upload_part(unsigned part_number, memory_data_sink_buffers bufs);
+    future<> abort_upload();
+
+    bool upload_started() const noexcept {
+        return !_upload_id.empty();
+    }
+
+public:
+    upload_sink(shared_ptr<client> cln, sstring object_name)
+        : _client(std::move(cln))
+        , _http(_client->_http)
+        , _object_name(std::move(object_name))
+    {
+    }
+
+    virtual future<> put(net::packet) override {
+        throw_with_backtrace<std::runtime_error>("s3 put(net::packet) unsupported");
+    }
+
+    virtual future<> put(temporary_buffer<char> buf) override {
+        _bufs.put(std::move(buf));
+        return maybe_flush();
+    }
+
+    virtual future<> put(std::vector<temporary_buffer<char>> data) override {
+        for (auto&& buf : data) {
+            _bufs.put(std::move(buf));
+        }
+        return maybe_flush();
+    }
+
+    virtual future<> flush() override {
```

Comment:
> If I understand correctly, contrary to regular sinks, you should not call flush() until the very end.
> Doing otherwise will cause the sink to be terminated and you won't be able to write more into it.

Correct

> Or maybe I misunderstood the interface and you are never supposed to call flush() in the middle in general?

There's no such requirement, so any writer is free to flush as often as it wants. And s3 sink will get broken after this, I didn't put any fuse against that.

Finalization of upload happening in flush() is not to do it in close(), because it apparently may fail, and failing close() was learned to be bad (@bhalevy must have scars left from those battles)

I planned to fix it together with #13019. Every flush would finalize temporary object and would then copy that object at the tail of the target. Provided current writers only flush once, this will kill two birds with one stone -- it will make flush() callable multiple times and won't extra hurt performance on real workload.